src/lib/operators/table_scan/sorted_segment_search.hpp

```diff
@@ -0,0 +1,165 @@
+        case PredicateCondition::LessThan: _begin = _get_last_bound(); return;
+        case PredicateCondition::GreaterThanEquals: _end = _get_last_bound(); return;
+        case PredicateCondition::GreaterThan: _end = _get_first_bound(); return;
+        default: Fail("Unsupported predicate condition encountered");
+      }
+    }
+    // clang-format on
+  }
+
+  /*
+   * NotEquals may result in two matching ranges (one below and one above the search_value) and needs special handling.
+   * The function contains three early outs. These are all only for performance reasons and, if removed, would not
+   * change the functionality.
+   */
+  template <typename Functor>
+  void _handle_not_equals(const Functor& functor) {
+    const auto first_occurrence = _get_first_bound();
+    if (first_occurrence == _end) {
+      // If the value isn't found anywhere, call the functor on the whole range and skip the call to _get_last_bound().
+      functor(_begin, _end);
+      return;
+    }
+
+    const auto last_occurrence = _get_last_bound();
+    if (last_occurrence == _end) {
+      // If the last occurrence of the value is right at the end, call the functor from start to first occurrence and
+      // skip the need for boost::join().
+      functor(_begin, first_occurrence);
+      return;
+    }
+
+    if (first_occurrence == _begin) {
+      // If the first occurrence of the value is right at the start, call the functor from last occurrence to end and
+      // skip the need for boost::join().
+      functor(last_occurrence, _end);
+      return;
+    }
+
+    const auto range = boost::join(boost::make_iterator_range(_begin, first_occurrence),
+                                   boost::make_iterator_range(last_occurrence, _end));
+    functor(range.begin(), range.end());
+  }
+
+ public:
+  template <typename Functor>
+  void scan_sorted_segment(const Functor& functor) {
+    // decrease the effective sort range by excluding null values based on their ordering
+    if (_is_nulls_first) {
+      _begin = std::lower_bound(_begin, _end, false,
+                                [](const auto& segment_position, const auto& _) { return segment_position.is_null(); });
+    } else {
+      _end = std::lower_bound(_begin, _end, true,
+                              [](const auto& segment_position, const auto& _) { return !segment_position.is_null(); });
+    }
+
+    if (_predicate_condition == PredicateCondition::NotEquals) {
+      _handle_not_equals(functor);
+    } else {
+      _set_begin_and_end();
+      functor(_begin, _end);
+    }
+  }
+
+ private:
+  IteratorType _begin;
+  IteratorType _end;
+  const PredicateCondition _predicate_condition;
+  const SearchValueType _search_value;
+  const bool _is_ascending;
+  const bool _is_nulls_first;
+};
+
+}  // namespace detail
+
+template <typename IteratorType, typename SearchValueType, typename Functor>
+void scan_sorted_segment(IteratorType begin, IteratorType end, const OrderByMode& order_by,
```

Comment:
Minor, but we now have `_scan_sorted_segment` and `scan_sorted_segment`. I'd remove `scan_sorted_segment` and just let `_scan_sorted_segment` call the `SortedSegmentSearch` itself. Would also make more sense when reading `_scan_sorted_segment`